libraries/common/array.effekt

```diff
@@ -185,6 +228,340 @@ def sum(list: Array[Int]): Int = {
   acc
 }
 
+/// Checks whether two given are equal with respect to an equalty funciton defined on their element.
+/// 
+/// O(N)
+def equals[A](this: Array[A], other: Array[A]) { eq: (A, A) => Bool }: Bool = {
+  if (this.size != other.size) false
+  else {
+    var res = true
+    this.foreachIndex { (idx, elem) {label} =>
+      if (not(eq(elem, other.unsafeGet(idx)))) {
+        res = false
+        label.break()
+      }
+    }
+    res
+  }
+}
+
+/// Map a function `f` over the elements in a given array.
+/// Modifies the array in-place.
+///
+/// O(N)
+def map[A](arr: Array[A]) { f: A => A }: Unit = {
+  arr.foreachIndex { (idx, elem) =>
+    arr.unsafeSet(idx, f(arr.unsafeGet(idx)))
+  }
+}
+
+/// Map a function over elements.
+/// Allocates a new array of the same size.
+///
+/// O(N)
+def mapped[A, B](arr: Array[A]) { f: A => B }: Array[B] = {
+  val res = allocate(arr.size)
+  arr.foreachIndex { (idx, elem) =>
+    res.unsafeSet(idx, f(arr.unsafeGet(idx)))
+  }
+  res
+}
+
+/// Reverses the array in-place
+/// 
+/// O(N)
+def reverse[A](arr: Array[A]): Unit = {
+  val size = arr.size()
+  val mid = size / 2
+
+  // Swap elements from start and end, moving towards the middle
+  each(0, mid) { i =>
+    val temp = arr.unsafeGet(i)
+    val j = size - 1 - i
+    val other = arr.unsafeGet(j)
+    arr.unsafeSet(i, other)
+    arr.unsafeSet(j, temp)
+  }
+}
+/// Creates a new array with elements in reverse order.
+/// 
+/// O(N)
+def reversed[A](arr: Array[A]): Array[A] = {
+  val size = arr.size()
+  val result = allocate[A](size)
+  
+  // Copy elements in reverse order to new array
+  each(0, size) { i =>
+    val value = arr.unsafeGet(i)
+    result.unsafeSet(size - 1 - i, value)
+  }
+  result
+}
+
+/// Fold a list using `f`, starting from the left given a starting value.
+///
+/// O(N)
+def foldLeft[A, B](arr: Array[A], init: B) { f: (B, A) => B }: B = {
+  var acc = init;
+  arr.foreach { x => acc = f(acc, x) };
+  acc
+}
+
+/// Fold a list using `f`, starting from the right given a starting value.
+///
+/// O(N)
+def foldRight[A, B](arr: Array[A], init: B) { f: (A, B) => B }: B = {
+  var acc = init;
+  arr.reversed.foreach { x => acc = f(x, acc) };
+  acc
+}
+
+/// Filters a given array with respect to a given predicate such that a
+/// new array containing all the elements for which the predicate
+/// holds is allocated.
+/// 
+/// O(N)
+def filtered[A](arr: Array[A]) { pred: A => Bool }: Array[A] = {
+  val nMatches = arr.count { pred }
+  val res = allocate(nMatches)
+  var idx = 0
+  arr.foreach { x =>
+    if (pred(x)) {
+      res.unsafeSet(idx, x)
+      idx = idx + 1
+    }
+  }
+  res
+}
```

Comment:
The semantics here can be a bit surprising: `pred` will get called twice on each element -- but it can have side effects which will be then executed multiple times. 👀

(similarly for `partitioned`)

Reply:
Since I'm nitpicking here already 😇 

```suggestion
/// Filters a given array with respect to a given predicate such that a
/// new array containing all the elements for which the predicate
/// holds is allocated.
/// 
/// O(N)
def filtered[A](arr: Array[A]) { shouldKeep: A => Bool }: Array[A] = {
  val nMatches = arr.count { shouldKeep }
  val res = allocate(nMatches)
  var idx = 0
  arr.foreach { x =>
    if (shouldKeep(x)) {
      res.unsafeSet(idx, x)
      idx = idx + 1
    }
  }
  res
}
```

Reply:
I agree that shrinking or dynamically resizing is better than calling it twice.